### ISSUE TITLE
When manually adding a subscriber, Stripe Source ID with "pm_" not working.

### ISSUE THREAD

**Issue by radiolize** [woocommerce/woocommerce-gateway-stripe#1181]
**Describe the bug**

Stripe using "pm_" for a long time. And we are trying to migrate our current customers to Woocommerce Subscription. When creating a subscriber manually at woocommerce subscription, this error appears.

**To Reproduce**
1. Go to woocommerce subscribers
2. Add manually subscriber
3. Insert your current customer's payment source starting with "pm_"
4. You will see "Error updating some information: Invalid source ID. A valid source "Stripe Source ID" must begin with "src_" or "card_"."

**Expected behavior**

Subscriber should be succesfully created.

**Environment (please complete the following information):**
 
Latest wordpress and woocommerce version.

**Additional context**

Waiting it for 1 year. Please let us know when it happen or if it does not so we will be looking another solutions.

**Comment by nsandrew** [comment]
We're running into the same bug trying to migrate clients from Squarespace subscriptions to WC + Subscriptions. Platforms that have fully migrated away from the Stripe Tokens and Sources APIs (which were replaced by the Payment Methods API) only have payment method ("pm_") IDs in their Stripe customer data. For context:
https://stripe.com/docs/payments/payment-methods/transitioning
especially
https://stripe.com/docs/payments/payment-methods/transitioning#compatibility

Recommend higher priority for this since it's a roadblock for migrating subscription data relying on Stripe payments from other platforms into WC, and Stripe has been moving away from the Tokens and Sources APIs for a while now. 

The problem is bigger than just the validation here:
https://github.com/woocommerce/woocommerce-gateway-stripe/blob/19e5f4a47d4ad94d4d367a588b3cd4b227c9f80d/includes/compat/class-wc-stripe-subs-compat.php#L507

The problem is that even though WC Stripe is calling the Payment Intents API it's passing the `source` parameter. So, even if a pm_##### ID was allowed in `_stripe_source_id` the request would still be wrong and cause an error 400 response from Stripe. The Payment Intents API still honors 'card_' and 'src_' tokens which has kept the pressure off full transition. 

The great news is that it also looks like the Payment Intents / Payment Methods API might also support original 'card_' and 'src_' tokens in the latest proper `payment_method` parameter as seen at the Stripe docs link above. That might mean that it could (hopefully) be as simple as:

1) Updating validation on subs-compat and sepa-subs-compat to allow 'pm_' prefix.

2) Updating payment requests to use the `payment_method` parameter everywhere instead of the `source` parameter. For example in `prepare_source()` here: 
https://github.com/woocommerce/woocommerce-gateway-stripe/blob/19e5f4a47d4ad94d4d367a588b3cd4b227c9f80d/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L591
and in `prepare_order_source()` here:
https://github.com/woocommerce/woocommerce-gateway-stripe/blob/19e5f4a47d4ad94d4d367a588b3cd4b227c9f80d/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L690

Here is a visual example of the output between the two requests for what is described above.
![image](https://user-images.githubusercontent.com/6741979/84351436-ff8fd000-ab80-11ea-9ef1-a39154e004c1.png)

We're going to take a stab at seeing if that approach with replacing the parameter name is all that's needed, and if we have any big insights we'll share! The powerful thing about getting this bug fixed is that as long as the subscription data has the Stripe 'cus_' ID and the 'pm_' ID from a previous system, the customer doesn't need to re-add their payment method in WC (until it expires) which dramatically improves retention by creating a no-friction migration for businesses that switch to WC from a different platform.

**Comment by nsandrew** [comment]
Looks like we were originally looking too far upstream and `generate_create_intent_request()` is probably a much better place to jump in.

Looks like it should even be possible to tweak the request by using the `wc_stripe_generate_create_intent_request` filter before the final request actually fires to Stripe, but we're struggling and maybe just don't have enough experience with the full cycle. For every 1 action (like a manual renewal order payment process action) we're seeing multiple (~5) requests getting sent in the gateway debug log. Is that normal?

If someone had a quick, working use of the `wc_stripe_generate_create_intent_request` filter to remove the `source` parameter and add the payment_method` parameter (set to the source ID value), we'd be ever grateful. Still hammering on it though.

**Comment by gppuddinpie** [comment]
nsandrew, I'm not a developer, however, this issue has kept us from migrating to subscriptions. I'd like to try modifying the code until they finally fix this issue. No clue why it's not higher priority. I'm sure it keeps everyone who uses Stripe on a different subscription platform from migrating. Have you validated that this works when manually adding subscriptions?

**Comment by nsandrew** [comment]
Hey, @gppuddinpie - yes, I can confirm those two code edits I described are working for us. The site has been live since that post (~3 weeks ago) and renewals are happening successfully. Migration can be really tricky though. We also discovered (unrelated to this) that some features for migrated subscriptions (like a customer being able to modify the subscription in their My Account area) might not work or be available if the subscription does NOT have a proper connection to a previous Renewal Order and the Last Order Date set. If you don't really support self-serve subscription modifications by your customers then this shouldn't really matter. Renewal billing will work with any of the following passed to Stripe in the `payment_method` parameter rather than the `source` parameter:
- src_
- card_
- pm_

**Comment by gppuddinpie** [comment]
@nsandrew , you're a genius. I implemented your suggestions and it worked like a charm in our test environment. Will have to look out for these other issues in the my-account area.

**Comment by icraftzone** [comment]
@nsandrew I am facing the same situation and tried your solutions but unfortunately, it's not working.

Trying to renew manually from the backend but getting this error: Error updating some information: Invalid source ID. A valid source "Stripe Source ID" must begin with "src_" or "card_".     even after commented out that validation check.

Can you please give me some light on this!

**Comment by icraftzone** [comment]
@gppuddinpie Can you please share how did you solve this? I am trying but seems nothing is working

**Comment by gppuddinpie** [comment]
I just commented out these lines:

woocommerce-gateway-stripe/includes/compat/class-wc-stripe-subs-compat.php
Line 507

`throw new Exception( __( 'Invalid source ID. A valid source "Stripe Source ID" must begin with "src_" or "card_".', 'woocommerce-gateway-stripe' ) ); `

And then modified the following file:
woocommerce-gateway-stripe/includes/abstracts/abstract-wc-stripe-payment-gateway.php

Lines 1379 to 1381:

```
 if ( isset( $full_request['source'] ) ) { 
 	$request['source'] = $full_request['source']; 
 }

```
modified like this:

```
if ( isset( $full_request['source'] ) ) {
	$request['source'] = $full_request['source'];
	unset( $request['source'] );
	$request['payment_method'] = $full_request['source'];
}
```

**Comment by davideprevosto** [comment]
Hello @gppuddinpie 

we are using the 4.5.0 version of WooCommerce Stripe Gateway plugin.

We have made changes to:

wp-content/plugins/woocommerce-gateway-stripe/includes/compat/class-wc-stripe-subs-compat.php:524 (commented the throw Exception)

wp-content/plugins/woocommerce-gateway-stripe/includes/abstracts/abstract-wc-stripe-payment-gateway.php:1427
```
if (isset($full_request['source'])) {
            $request['source'] = $full_request['source'];
            // begin @stripefix
            unset($request['source']);
            $request['payment_method'] = $full_request['source'];
            // end @stripefix
        }
```

It seems we are able to import subscription from our MemberPress legacy to WooCommerce. I have yet to understand changes in the woocommerce token table... because I am not able to store a source stripe id here, maybe it is no more useful. @nsandrew maybe do you notice any problem with src_ token and table wp_woocommerce_payment_tokens / wp_woocommerce_payment_tokenmeta?

Thank you

**Comment by nsandrew** [comment]
Sorry, @davideprevosto I'm not very familiar with the MemberPress data structure. However, the WC payment token data in those tables almost doesn't matter. What you should be looking for is a Stripe payment_method ID prefixed with one of:
src_
card_
pm_
That payment_method data is stored in the subscription postmeta. Your import process from [some system like MemberPress] into WC Subscriptions should be creating postmeta for the payment_method key. You might have to look at how WC Subscriptions stores that data on an example subscription created in WC and then make sure your import of data creates the correct key. The code updates above won't matter if you are not importing payment_method data per Subscription. You can simulate / edit this field in an existing subscription in the Billing Address of the subscription order under the address details it will show the active payment method. That needs to have a type and a value because that is what WC Subscriptions will send to the Stripe API for renewal orders. If there's no value there or you don't have those values to import from the old system then it won't work. However, you could extract data from Stripe, match on customer email, and get those payment_method(s) per customer for your import data. But you should test this in a very small batch first. 

Bottom line is - for the purposes of importing and getting this hack to work - you can completely ignore the tokens and token tables. They won't help you in this case.

**Comment by gppuddinpie** [comment]
@nsandrew , will the prefix of pm_ be implemented into subscriptions any time soon? As far as I can see, without the override, it still only allows src_ and card_

**Comment by nsandrew** [comment]
I have no idea @gppuddinpie - I'm not on the WC team :) It's been marked as a low priority bug, so it all depends on when it gets into an official release. The work-around will indeed be needed until then.

**Comment by gppuddinpie** [comment]
My bad. Sorry, thought you were :)

**Comment by radiolize** [comment]
Because of this issue, woocommerce subscription plugin probably lost a lot of active businesses such as us for last 9 months as people are still not able to migrate their current subscribers for **9 long months**. The plugin's users -who paid 200 usd for it- are trying to find a solution themselves for months instead of well-known woocommerce's developers or product leads and still no news.

Can you guys tell your customers such as us when it will be done as **we are tired of seeing delayed milestones?** Should we wait forever or we are about to find another solution instead of woocommerce subscription plugin.

Thanks.

**Comment by scottbuscemi** [comment]
@radiolize The fix provided by @gppuddinpie above patches the issue. The WooCommerce team is aware of this and has told me there's something in the works to resolve this soon.
